src/util/internalCallTree.js

```diff
@@ -0,0 +1,174 @@
+    this.includeLocalVariables = opts.includeLocalVariables
+    this.event = new EventManager()
+    this.solidityProxy = solidityProxy
+    this.traceManager = traceManager
+    this.sourceLocationTracker = new SourceLocationTracker(codeManager)
+    debuggerEvent.register('newTraceLoaded', (trace) => {
+      this.reset()
+      if (!this.solidityProxy.loaded()) {
+        console.log('compilation result not loaded. Cannot build internal call tree')
+      } else {
+        buildTree(this, 0, '').then((result) => {
+          if (result.error) {
+            this.event.trigger('callTreeBuildFailed', [result])
+          } else {
+            console.log('ready')
+            this.event.trigger('callTreeReady', [this.scopes, this.scopeStarts])
+          }
+        })
+      }
+    })
+  }
+
+  /**
+    * reset tree
+    *
+    */
+  reset () {
+    /*
+      scopes: map of scopes defined by range in the vmtrace {firstStep, lastStep, locals}. Keys represent the level of deepness (scopeId)
+    */
+    this.scopes = {}
+    /*
+      scopeStart: represent start of a new scope. Keys are index in the vmtrace, values are scopeId
+    */
+    this.scopeStarts = {}
+    this.variableDeclarationByFile = {}
+    this.astWalker = new AstWalker()
+  }
+
+  /**
+    * find the scope given @arg vmTraceIndex
+    *
+    * @param {Int} vmtraceIndex  - index on the vm trace
+    */
+  findScope (vmtraceIndex) {
+    var scopes = Object.keys(this.scopeStarts)
+    if (!scopes.length) {
+      return null
+    }
+    var scopeId = util.findLowerBoundValue(vmtraceIndex, scopes)
+    scopeId = this.scopeStarts[scopeId]
+    var scope = this.scopes[scopeId]
+    var reg = /(.\d|\d)$/
+    while (scope.lastStep && scope.lastStep < vmtraceIndex) {
+      var matched = scopeId.match(reg)
+      scopeId = scopeId.replace(matched[1], '')
+      scope = this.scopes[scopeId]
+    }
+    return scope
+  }
+}
+
+async function buildTree (tree, step, scopeId) {
+  let subScope = 1
+  tree.scopeStarts[step] = scopeId
+  tree.scopes[scopeId] = { firstStep: step, locals: {} }
+  while (step < tree.traceManager.trace.length) {
+    var sourceLocation
+    try {
+      sourceLocation = await extractSourceLocation(tree, step)
+    } catch (e) {
+      return { outStep: step, error: 'InternalCallTree - Error resolving source location. ' + step + ' ' + e.messager }
+    }
+    if (!sourceLocation) {
+      return { outStep: step, error: 'InternalCallTree - No source Location. ' + step }
+    }
+    if (sourceLocation.jump === 'i') {
+      var result = await buildTree(tree, step + 1, scopeId === '' ? subScope.toString() : scopeId + '.' + subScope)
+      if (result.error) {
+        return result
+      } else {
+        step = result.outStep
+        subScope++
+      }
+    } else if (sourceLocation.jump === 'o') {
+      tree.scopes[scopeId].lastStep = step
+      return { outStep: step + 1 }
+    } else {
+      if (tree.includeLocalVariables) {
```

Comment:
Turn into function (async reference to step will be invalidated in the next loop iteration).